kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue

```diff
@@ -286,15 +286,21 @@
       };
       this.eventBus = new EventBus();
       this.prepComponentData = loadingPdf.promise
-        .then(pdfDocument => {
+        .then(async pdfDocument => {
           // Get initial info from the loaded pdf document
           this.pdfDocument = pdfDocument;
           this.totalPages = this.pdfDocument.numPages;
           // init pdfPages array
           for (let i = 0; i < this.totalPages; i++) {
+            const pageSize = await this.pdfDocument.getPage(i + 1).then(page => {
```

Comment:
Here we are loading all pages of the document at once, at the first pdf load, this could bring performance issues at first load for all pdf documents, including those whose does not have different page heights (which are the majority), so I would not introduce this additional load on performance just to correct the behavior of a minority of documents. I would suggest to calculate this first item size with the height of the first rendered page, and then when we load new pages, we can update the `size` method:

```JS
          for (let i = 0; i < this.totalPages; i++) {
            this.pdfPages.push({
              page: null,
              resolved: false,
              size: () => {
                return this.itemHeight; // height of the first page to be rendered
                // just saw that the height is calculated below this code, but the idea remains the same, set the first page height for all pages at first load
              },
              index: i,
          }
```

And when the pages [are loaded](https://github.com/ThEditor/kolibri/blob/3b6477672cb4b4339f0c7a600aa44d2eec96c3ae/kolibri/plugins/pdf_viewer/assets/src/views/PdfRendererIndex.vue#L383) we can update the function:

```JS
      showPage(pageNum) {
        // Only bother getting it if the pdfPage object is not already cached in the array
        if (pageNum > 0 && pageNum <= this.totalPages && !this.pdfPages[pageNum - 1].resolved) {
          const pageIndex = pageNum - 1;
          this.getPage(pageNum).then(pdfPage => {
            const { height } = pdfPage.getViewport({ scale: 1 });

            // splice so changes are detected
            this.pdfPages.splice(pageIndex, 1, {
              ...this.pdfPages[pageIndex],
              page: pdfPage,
              resolved: true,
              size: () => {
                return height * this.scale + MARGIN;
              },
            });
          });
        }
      },
```
But this approach can lead to "jumps" in the scroll when pages with different height are loaded:

https://github.com/learningequality/kolibri/assets/51239030/41e970a8-b4f4-4796-9d03-c106c1150516

@rtibbles What are your thoughts on this?

Reply:
Since we are rendering the pages one after the other, I think we can modify RecycleScroller further to load the next/previous page's size dynamically, I'll try to give my idea a shot and see if it works. I'll keep you posted.

Reply:
Hi @ThEditor! I talked with Richard, and he agreed on allow the slight jump, and we can after open a follow up issue to try to ameliorate it. But for now we can let the size be redefined when the pages load on demand as I suggested above :open_hands: